core/trino-main/src/main/java/io/trino/operator/MultiChannelGroupByHash.java

```diff
@@ -980,55 +990,54 @@ public GroupByIdBlock getResult()
         }
     }
 
-    private abstract class LowCardinalityDictionaryWork<T>
-            implements Work<T>
+    /**
+     * Returns an array containing a single Page position occurrence that corresponds to the
+     * low cardinality dictionary combinationId at that index, or a value of -1 if no position
+     * within exists within the page for that combinationId.
+     */
+    private int[] calculateCombinationIdToPositionMapping(Page page)
     {
-        protected final Page page;
-        protected final int maxCardinality;
-        protected final int[] dictionarySizes;
-        protected final DictionaryBlock[] blocks;
+        short[] positionToCombinationIds = new short[page.getPositionCount()];
+        int maxCardinality = calculatePositionToCombinationIdMapping(page, positionToCombinationIds);
 
-        public LowCardinalityDictionaryWork(Page page)
-        {
-            this.page = requireNonNull(page, "page is null");
-            dictionarySizes = new int[channels.length];
-            blocks = new DictionaryBlock[channels.length];
-            int maxCardinality = 1;
-            for (int i = 0; i < channels.length; i++) {
-                Block block = page.getBlock(channels[i]);
-                verify(block instanceof DictionaryBlock, "Only dictionary blocks are supported");
-                blocks[i] = (DictionaryBlock) block;
-                int blockPositionCount = blocks[i].getDictionary().getPositionCount();
-                dictionarySizes[i] = blockPositionCount;
-                maxCardinality *= blockPositionCount;
-            }
-            this.maxCardinality = maxCardinality;
+        int[] combinationIdToPosition = new int[maxCardinality];
+        Arrays.fill(combinationIdToPosition, -1);
+        for (int position = 0; position < positionToCombinationIds.length; position++) {
+            combinationIdToPosition[positionToCombinationIds[position]] = position;
         }
+        return combinationIdToPosition;
+    }
 
-        /**
-         * Returns combinations of all dictionaries ids for every position and populates
-         * samplePositions array with a single occurrence of every used combination
-         */
-        protected short[] calculateCombinationIdsToPositionMapping(int[] combinationIdToPosition)
-        {
-            int positionCount = page.getPositionCount();
-            // short arrays improve performance compared to int
-            short[] combinationIds = new short[positionCount];
-
-            for (int i = 0; i < positionCount; i++) {
-                combinationIds[i] = (short) blocks[0].getId(i);
-            }
-            for (int j = 1; j < channels.length; j++) {
-                for (int i = 0; i < positionCount; i++) {
-                    combinationIds[i] *= dictionarySizes[j];
-                    combinationIds[i] += blocks[j].getId(i);
+    /**
+     * Returns the number of combinations of all dictionaries ids for every position and populates
```

Comment:
`Returns the number of combinations of all dictionaries ids for every position` -> `Returns the number of combinations of all dictionaries ids in input page blocks`